bundles/org.openhab.core.config.discovery.addon.ip/src/main/java/org/openhab/core/config/discovery/addon/ip/IpAddonFinder.java

```diff
@@ -0,0 +1,265 @@
+                    continue;
+                }
+                @Nullable
+                InetAddress destIp = null;
+                try {
+                    destIp = InetAddress.getByName(parameters.get("dest_ip"));
+                } catch (UnknownHostException e) {
+                    logger.info("{}: discovery-parameter dest_ip cannot be parsed", candidate.getUID());
+                    continue;
+                }
+                int destPort = 0;
+                try {
+                    destPort = Integer.parseInt(Objects.toString(parameters.get("dest_port")));
+                } catch (NumberFormatException e) {
+                    logger.info("{}: discovery-parameter dest_port cannot be parsed", candidate.getUID());
+                    continue;
+                }
+
+                //
+                // handle known types
+                //
+                try {
+                    switch (Objects.toString(type)) {
+                        case "ip_multicast":
+                            List<String> ipAddresses = NetUtil.getAllInterfaceAddresses().stream()
+                                    .filter(a -> a.getAddress() instanceof Inet4Address)
+                                    .map(a -> a.getAddress().getHostAddress()).toList();
+
+                            for (String localIp : ipAddresses) {
+                                try {
+                                    DatagramChannel channel = (DatagramChannel) DatagramChannel
+                                            .open(StandardProtocolFamily.INET)
+                                            .setOption(StandardSocketOptions.SO_REUSEADDR, true)
+                                            .bind(new InetSocketAddress(localIp, 0))
+                                            .setOption(StandardSocketOptions.IP_MULTICAST_TTL, 64)
+                                            .configureBlocking(false);
+
+                                    byte[] requestArray = buildRequestArray(channel, Objects.toString(request));
+                                    logger.trace("{}: {}", candidate.getUID(),
+                                            HexFormat.of().withDelimiter(" ").formatHex(requestArray));
+
+                                    channel.send(ByteBuffer.wrap(requestArray),
+                                            new InetSocketAddress(destIp, destPort));
+
+                                    // listen to responses
+                                    Selector selector = Selector.open();
+                                    ByteBuffer buffer = ByteBuffer.wrap(new byte[50]);
+                                    channel.register(selector, SelectionKey.OP_READ);
+                                    selector.select(timeoutMs);
+                                    Iterator<SelectionKey> it = selector.selectedKeys().iterator();
+
+                                    switch (Objects.toString(response)) {
+                                        case ".*":
```

Comment:
I don't think this can be generalized. Can't you read the buffer, convert to String and do a regex on it?
We could leave this improvement for later when we make more add-ons discoverable through this finder. But it should be improved.

Reply:
yes, in principle....

Reply:
The long answer to this is.....
Yes of course we can hex dump the received frame and do a regex match. Maybe the regex needs to be dynamic (using $srcIp etc before doing the match).
I would need to change the receive loop, as just taking the first frame and return is no longer possible. I.e. a receive loop, which takes care for the overall timeout and then checking every frame within the time period.

Not sure if this is required for the first implementation, I would probably leave a shortcut for ".*" in anyway as it avoids the regex matching etc.

Reply:
I am OK to improve this for 4.2. I hope you will do some work to look at all bindings that use this type of discovery and try to cover more of these. Start by defining criteria, and then see where this core matching needs to ba adapted.